### PR TITLE
3D LR projection and bezier evaluation

### PR DESCRIPTION
**Pull request by VikingScientist**
This is a remake of PR #52 which had to be taken down for reasons described there.

### CODE REVIEW

File: src/ASM/LR/ASMu2D.C [hunk 1]
```diff
@@ -861,21 +861,20 @@ size_t ASMu2D::getNoBoundaryElms (char lIndex, char ldim) const
...
-    case 2: if ((*el)->umax() == lrspline->endparam(0))   nbel++; break;
-    case 3: if ((*el)->vmin() == lrspline->startparam(1)) nbel++; break;
-    case 4: if ((*el)->vmax() == lrspline->endparam(1))   nbel++; break;
-    }
+  case 1: edge = LR::WEST;
+  case 2: edge = LR::EAST;
+  case 3: edge = LR::SOUTH;
+  case 4: edge = LR::NORTH;
+  default:edge = LR::NONE;
   }
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2D.C]: Du her har mista break; i hver case her. Slik det står nå vil den gåt til default edge = LR::NONE uansett verdi for lIndex. Det var vel ikke hensikten?

File: src/ASM/LR/ASMu3D.C [hunk 2]
```diff
@@ -744,36 +744,23 @@ bool ASMu3D::updateCoords (const Vector& displ)
...
+	switch(lIndex)
+	{
+	case 1: edge = LR::WEST;
+	case 2: edge = LR::EAST;
+	case 3: edge = LR::SOUTH;
+	case 4: edge = LR::NORTH;
+	case 5: edge = LR::BOTTOM;
+	case 6: edge = LR::TOP;
+	default:edge = LR::NONE;
+	}
```
Comment (kmokstad) [inline src/ASM/LR/ASMu3D.C]: Her også mangler du break;

**Comment by kmokstad** [comment]
The first two commits in this one sit also in #69 which already has been merged, så please rebase this one against master before progressing, to avoid confusion for reviewers. #69 one should also have been rebased _before_ that merge...

**Comment by VikingScientist** [comment]
Updated requested changes. Also, this fixes issue #76.

I know tab-indentation should be fixed, but this will be treated in a "tab-fix-only" pull-request at a later point.

**Comment by VikingScientist** [comment]
Fixed comments made by @akva2 in person... Tests will appear in the Poisson repository. Ready for merging.

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-poisson=5 please

**Comment by VikingScientist** [comment]
All tests passed... with your permission, I would like to autosquash and make ready for merge now.

**Comment by akva2** [comment]
do it

File: src/ASM/LR/ASMu3D.C [hunk 3]
```diff
@@ -1796,14 +1814,18 @@ bool ASMu3D::evalSolution (Matrix& sField, const Vector& locSol,
 bool ASMu3D::evalSolution (Matrix& sField, const IntegrandBase& integrand,
                            const int* npe, char project) const
 {
+  if (npe != nullptr && npe[0] != npe[1] && npe[0] != npe[2] && npe[1] != npe[2])
```
Comment (akva2) [inline src/ASM/LR/ASMu3D.C]: code does not correspond to error message. i think you mean
```c++
if (npe != nullptr && (npe[0] != npe[1] || npe[0] != npe[2] || npe[1] != npe[2]))

File: src/ASM/LR/ASMu3Drecovery.C [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+  {
+#if SP_DEBUG > 2
+    std::cout <<"Basis: "<< *b <<"\n  ng1 ="<< ng1 <<"\n  ng2 ="<< ng2 <<"\n  ng3 ="<< ng3
+              <<"\n  nPol="<< nPol << std::endl;
+#endif
+
+    // Special case for basis functions with too many zero knot spans by using
+    // the extended support
+    // if(nel*ng1*ng2*ng3 < nPol)
+    if(true)
```
Comment (akva2) [inline src/ASM/LR/ASMu3Drecovery.C]: some comment about the if(true) here is appropriate (as discussed yesterday).

**Comment by VikingScientist** [comment]
Anything holding back the "merge" button?

**Comment by notspiff** [comment]
i'm waiting for you to fix the issue i pointed out above.

**Comment by akva2** [comment]
^^ me

**Comment by VikingScientist** [comment]
Fixed

**Comment by akva2** [comment]
lies.

**Pull request merged**